### PR TITLE
ASoC: SOF: sof-audio: Set up widgets from source to sink

### PR DESCRIPTION
**Pull request by ranj063**
For IPC3, the order of setting up the widgets associated with a PCM
doesn't matter. But for IPC4, widgets must be set up from the source to
the sink in order. In order to accommodate this, change the
sof_widget_list_setup/free() functions to set up/free widgets starting
with the source widget all the way to the sink widget for all pipelines.

Signed-off-by: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>

### CODE REVIEW

File: sound/soc/sof/sof-audio.c [hunk 1]
```diff
@@ -256,29 +256,125 @@ static int sof_setup_pipeline_connections(struct snd_sof_dev *sdev,
 	return 0;
 }
 
+/*
+ * free all widgets in the sink path starting from the source widget
+ * (DAI type for capture, AIF type for platback)
```
Comment (yaochunhung) [inline sound/soc/sof/sof-audio.c]: typo for playback, thanks.

File: sound/soc/sof/sof-audio.c [hunk 2]
```diff
@@ -256,29 +256,125 @@ static int sof_setup_pipeline_connections(struct snd_sof_dev *sdev,
...
+				ret1 = ret;
+		}
+	}
+
+	return ret1;
+}
+
+/*
+ * set up all widgets in the sink path starting from the source widget
+ * (DAI type for capture, AIF type for platback).
```
Comment (yaochunhung) [inline sound/soc/sof/sof-audio.c]: ditto

**Review by bardliao — APPROVED** [review]

File: sound/soc/sof/sof-audio.c [hunk 3]
```diff
@@ -256,29 +256,125 @@ static int sof_setup_pipeline_connections(struct snd_sof_dev *sdev,
...
+/*
+ * free all widgets in the sink path starting from the source widget
+ * (DAI type for capture, AIF type for playback)
+ */
+static int sof_free_widgets_in_path(struct snd_sof_dev *sdev, struct snd_soc_dapm_widget *widget,
+				    int dir)
+{
+	struct snd_soc_dapm_path *p;
+	int ret;
+	int ret1 = 0;
```
Comment (kv2019i) [inline sound/soc/sof/sof-audio.c]: I find this a bit hard to follow. I know we've used same approach e.g. in hda-dsp.c to have throwaway return variable, but e.g. in hda-dsp it's "ret1" that is the throwaway and "ret" is what we return up. Here it's the reverse.

Maybe name these "ret0" and "ret1" and have "ret0" as the outer result?  Dunno, no strong preference, just noting I find this hard to follow when reviewing.

File: sound/soc/sof/sof-audio.c [hunk 4]
```diff
@@ -256,29 +256,125 @@ static int sof_setup_pipeline_connections(struct snd_sof_dev *sdev,
...
+			}
+		}
+	}
+
+	return 0;
+}
+
+static int
+sof_setup_or_free_widgets_in_order(struct snd_sof_dev *sdev, struct snd_soc_dapm_widget_list *list,
+				   bool dir, bool setup)
```
Comment (kv2019i) [inline sound/soc/sof/sof-audio.c]: I know this is not specific to this comment, but can you please add a define or enum for 'setup'. It would be much easier to read the code if it states "sof_state_or_free_widgets_in_oder(sdev, list, PLAYBACK, SETUP)" than "sof_state_or_free_widgets_in_oder(sdev, list, PLAYBACK, true)"

File: sound/soc/sof/sof-audio.c [hunk 5]
```diff
@@ -256,29 +256,125 @@ static int sof_setup_pipeline_connections(struct snd_sof_dev *sdev,
...
+			ret = sof_set_up_widgets_in_path(sdev, widget, dir);
+			if (ret < 0) {
+				dev_err(sdev->dev, "Failed to set up connected widgets\n");
+				return ret;
+			}
+		} else {
+			ret = sof_free_widgets_in_path(sdev, widget, dir);
+			if (ret < 0) {
+				dev_err(sdev->dev, "Failed to free connected widgets\n");
+				return ret;
```
Comment (kv2019i) [inline sound/soc/sof/sof-audio.c]: You can simply this by having a single check for ret and print:
dev_err(sdev->dev, "Failed to %s connected widgets\n"), setup ? "setup" : "free");

File: sound/soc/sof/sof-audio.c [hunk 6]
```diff
@@ -256,29 +256,125 @@ static int sof_setup_pipeline_connections(struct snd_sof_dev *sdev,
...
+ * (DAI type for capture, AIF type for playback).
+ * The error path in this function ensures that all successfully set up widgets getting freed.
+ */
+static int sof_set_up_widgets_in_path(struct snd_sof_dev *sdev, struct snd_soc_dapm_widget *widget,
+				      int dir)
+{
+	struct snd_soc_dapm_path *p;
+	int ret;
+
+	snd_soc_dapm_widget_for_each_sink_path(widget, p) {
```
Comment (kv2019i) [inline sound/soc/sof/sof-audio.c:302]: Does this work for capture widget or should snd_soc_dapm_widget_for_each_source_path() be used (all paths out from the capture widget)?

**Review by kv2019i — CHANGES_REQUESTED** [review]
Commit description is easy to read and main logic seems while. One question on DAPM iteration and a few coding style notes.

File: sound/soc/sof/sof-audio.c (hunk 3 above)
Comment (ujfalusi) [reply sound/soc/sof/sof-audio.c]: I agree, at the end of the function it is much better to have `return ret;`

File: sound/soc/sof/sof-audio.c [hunk 7]
```diff
@@ -256,29 +256,125 @@ static int sof_setup_pipeline_connections(struct snd_sof_dev *sdev,
...
+			continue;
+
+		/* starting widget for capture is DAI type */
+		if (dir == SNDRV_PCM_STREAM_CAPTURE && !WIDGET_IS_DAI(widget->id))
+			continue;
+
+		if (setup) {
+			ret = sof_set_up_widgets_in_path(sdev, widget, dir);
+			if (ret < 0) {
+				dev_err(sdev->dev, "Failed to set up connected widgets\n");
```
Comment (ujfalusi) [inline sound/soc/sof/sof-audio.c]: print out the name of the widget as a hint?

File: sound/soc/sof/sof-audio.c [hunk 8]
```diff
@@ -256,29 +256,125 @@ static int sof_setup_pipeline_connections(struct snd_sof_dev *sdev,
...
+	int ret, i;
+
+	for_each_dapm_widgets(list, i, widget) {
+		/* starting widget for playback is AIF type */
+		if (dir == SNDRV_PCM_STREAM_PLAYBACK && !WIDGET_IS_AIF(widget->id))
+			continue;
+
+		/* starting widget for capture is DAI type */
+		if (dir == SNDRV_PCM_STREAM_CAPTURE && !WIDGET_IS_DAI(widget->id))
+			continue;
```
Comment (ujfalusi) [inline sound/soc/sof/sof-audio.c:348]: You are looking up the start of the path, right? Is it expected that you will have more than one start entry?
```c
struct snd_soc_dapm_widget *widget;
bool found = false;
int i;

for_each_dapm_widgets(list, i, widget) {
	if (dir == SNDRV_PCM_STREAM_PLAYBACK && WIDGET_IS_AIF(widget->id) ||
	    dir == SNDRV_PCM_STREAM_CAPTURE && WIDGET_IS_DAI(widget->id)) {
		found = true;
		break;
	}
}

if (found) {
	int ret;

	if (setup)
		ret = sof_set_up_widgets_in_path(sdev, widget, dir);
	else
		ret = sof_free_widgets_in_path(sdev, widget, dir);

	if (ret < 0) {
		dev_err(sdev->dev, "Failed to %s connected widgets\n",
			setup ? "set up" : "free");
		return ret;
	}
}

return 0;
```
Is it considered OK if we do not find the start?

**Review by ujfalusi — CHANGES_REQUESTED** [review]
We need to set up the widgets from start to end, make sense. Is it so that the same applies when tearing it down? We are not rewinding, but again starting from the root?

File: sound/soc/sof/sof-audio.c (hunk 7 above)
Comment (ranj063) [reply sound/soc/sof/sof-audio.c]: it will be printed inside the sof_widget_free() function for whichever widget that fails

File: sound/soc/sof/sof-audio.c (hunk 8 above)
Comment (ranj063) [reply sound/soc/sof/sof-audio.c:348]: Yes, there can be more than one starting widget. I cannot quite envision this case for playback but for capture it is certainly a possibility.

File: sound/soc/sof/sof-audio.c (hunk 6 above)
Comment (ranj063) [reply sound/soc/sof/sof-audio.c:302]: @kv2019i yes, this works for capture as well. Basically I only traverse the sink path starting from the source widget in both cases. For playback it will be the AIF widget and DAI widget for capture. The reason for this is that it is easy to unroll and free previously set up widgets if something fails along the way

File: sound/soc/sof/sof-audio.c (hunk 8 above)
Comment (plbossart) [reply sound/soc/sof/sof-audio.c:348]: beep tone generated in firmware and played on multiple outputs...

Comment (ranj063) [reply sound/soc/sof/sof-audio.c:348]: > beep tone generated in firmware and played on multiple outputs...

But in this case the host widget will be only one still right?

**Review by kv2019i — APPROVED** [review]

File: sound/soc/sof/sof-audio.c [hunk 9]
```diff
@@ -256,29 +256,128 @@ static int sof_setup_pipeline_connections(struct snd_sof_dev *sdev,
...
+				err = sof_widget_free(sdev, widget->dobj.private);
+				if (err < 0)
+					ret = err;
+			}
+
+			err = sof_widget_free(sdev, p->sink->dobj.private);
+			if (err < 0)
+				ret = err;
+
+			err = sof_free_widgets_in_path(sdev, p->sink, dir);
```
Comment (plbossart) [inline sound/soc/sof/sof-audio.c:281]: what happens if there is a loop in the dapm graph?

File: sound/soc/sof/sof-audio.c [hunk 10]
```diff
@@ -256,29 +256,128 @@ static int sof_setup_pipeline_connections(struct snd_sof_dev *sdev,
...
+		if (p->sink->dobj.private && widget->dobj.private) {
+			if (WIDGET_IS_AIF(widget->id) || WIDGET_IS_DAI(widget->id)) {
+				ret = sof_widget_setup(sdev, widget->dobj.private);
+				if (ret < 0)
+					return ret;
+			}
+
+			ret = sof_widget_setup(sdev, p->sink->dobj.private);
+			if (ret < 0) {
+				sof_widget_free(sdev, widget->dobj.private);
```
Comment (plbossart) [inline sound/soc/sof/sof-audio.c]: is this intentional? You don't use the free line 308 if sof_widget_setup() fails.

Comment (ranj063) [reply sound/soc/sof/sof-audio.c]: oh, so we set up 2 widgets for each route, the source (only if it is a DAI/AIF widget) and the sink widget. If the source is successful but the sink set up fails, we need to free the source. But I missed the if (WIDGET_IS_AIF(widget->id) || WIDGET_IS_DAI(widget->id)) , will fix.

File: sound/soc/sof/sof-audio.c (hunk 9 above)
Comment (ranj063) [reply sound/soc/sof/sof-audio.c:281]: good point. let me add a condition to stop at the starting widget in case of a loop.

Comment (plbossart) [reply sound/soc/sof/sof-audio.c:281]: the loop could be just after the starting widget... you may need something that tracks all widgets to see if you've already handled them.

Wondering if there isn't already defined in ASoC/DAPM provides for 'safe' walks through a graph with loops?

Comment (ranj063) [reply sound/soc/sof/sof-audio.c:281]: @plbossart It was quite simple to prevent an infinite walk. we can mark the paths as walking or not as we traverse the list.

File: sound/soc/sof/sof-audio.c [hunk 11]
```diff
@@ -256,29 +256,133 @@ static int sof_setup_pipeline_connections(struct snd_sof_dev *sdev,
...
+	struct snd_soc_dapm_path *p;
+	int ret;
+
+	snd_soc_dapm_widget_for_each_sink_path(widget, p) {
+		if (!p->walking && p->sink->dobj.private && widget->dobj.private) {
+			p->walking = true;
+			if (WIDGET_IS_AIF_OR_DAI(widget->id)) {
+				ret = sof_widget_setup(sdev, widget->dobj.private);
+				if (ret < 0)
+					return ret;
```
Comment (plbossart) [inline sound/soc/sof/sof-audio.c]: should p->walking be marked as false in case of errors?

File: sound/soc/sof/sof-audio.c [hunk 12]
```diff
@@ -256,29 +256,133 @@ static int sof_setup_pipeline_connections(struct snd_sof_dev *sdev,
...
+			}
+
+			ret = sof_widget_setup(sdev, p->sink->dobj.private);
```
Comment (plbossart) [inline sound/soc/sof/sof-audio.c:311]: in the case of IS_AIF_OR_DAI, we will have two calls to sof_widget_setup(), is this intentional?

File: sound/soc/sof/sof-audio.c [hunk 13]
```diff
@@ -256,29 +256,133 @@ static int sof_setup_pipeline_connections(struct snd_sof_dev *sdev,
...
+static int sof_free_widgets_in_path(struct snd_sof_dev *sdev, struct snd_soc_dapm_widget *widget,
+				    int dir)
+{
+	struct snd_soc_dapm_path *p;
+	int err;
+	int ret = 0;
+
+	/* free all widgets even in case of error to keep use counts balanced */
+	snd_soc_dapm_widget_for_each_sink_path(widget, p) {
+		if (!p->walking && p->sink->dobj.private && widget->dobj.private) {
```
Comment (plbossart) [inline sound/soc/sof/sof-audio.c:269]: how do we know that p->walking was properly reset? defensive programming would be nice with this field explicitly cleared before starting the recursion - maybe with a helper doing the recursion after clearing the field?

**Review by plbossart — COMMENTED** [review]
see questions below @ranj063, the code looks fine but as usual I am worried about error handling. There's also a possible double setup and double free that I didn't get at all.

File: sound/soc/sof/sof-audio.c (hunk 13 above)
Comment (ranj063) [reply sound/soc/sof/sof-audio.c:269]: I think it is a bit of a n overkill, Pierre. With p->walking set to false after every successful/failed walk, I dont think we need explicit clearing before.

File: sound/soc/sof/sof-audio.c (hunk 11 above)
Comment (ranj063) [reply sound/soc/sof/sof-audio.c]: yes, it should. fixed now.

File: sound/soc/sof/sof-audio.c (hunk 12 above)
Comment (ranj063) [reply sound/soc/sof/sof-audio.c:311]: yes, basically the starting widget and its sink should both be set up.

**Review by ujfalusi — DISMISSED** [review]

File: sound/soc/sof/sof-audio.c (hunk 13 above)
Comment (plbossart) [reply sound/soc/sof/sof-audio.c:269]: ok, I hope I won't have to say "I told you so" :-)

File: sound/soc/sof/sof-audio.c (hunk 12 above)
Comment (plbossart) [reply sound/soc/sof/sof-audio.c:311]: ah yes, read sideways and missed the p->sink.

**Review by plbossart — DISMISSED** [review]

**Comment by plbossart** [comment]
LGTM but will wait for @kv2019i @bardliao @yaochunhung to revisit the changes.

**Review by kv2019i — APPROVED** [review]

**Comment by ranj063** [comment]
@plbossart @ujfalusi @kv2019i fixed a cppcheck error. Please have a look again.

**Review by bardliao — APPROVED** [review]

**Review by ujfalusi — APPROVED** [review]
